### DISCUSSION TITLE
How to backup photos when it does not fit on my machine any more?

### DISCUSSION THREAD

**Discussion by synox** [RhetTbull/osxphotos#397] [Q&A]
With osxphotos I can finally safely export all my library into a structured way. 

Do I understand correctly that I need to have "Download Originals to this Mac" enabled, and not "Optimise Mac Storage"? 

My Photo libary is soon larger than my hard drive. It it feasible to keep exporting using "Optimise Mac Storage"?

**Comment by RhetTbull** [comment]
osxphotos can only export photos that are on the local disk as it works be directly interacting with the Photos library database.  In your case, if using "Optimize Mac Storage", if you run osxphotos you'll see that a lot of photos are "missing" and don't get exported because they're not on the local Mac.  osxphotos provides the `--download-missing` option to force Photos to download these to the local Mac.  By itself, `--download-missing` uses AppleScript to interact with Photos.  There's also a `--use-photokit` option that can be used with `--download-missing` to use a lower-level interface to Photos and get the photos from iCloud.  I don't know what Photos would do if you run out of disk space during the export process.  It might continually optimize as osxphotos is forcing it to download thus deleting the originals from the Mac as osxphotos runs or it might just fill up the disk.  It depends on whether Photos does this continually in the background or if it's a batch process that runs at certain times.  

Another option in your case is to use [iCloud Photos Downloader](https://github.com/icloud-photos-downloader/icloud_photos_downloader), another project who's purpose is to simply download photos off iCloud.  In your case you could have those downloaded to external disk.  I've never used this app so I can't vouch for it but it seems popular and well-maintained.

Finally, you could also move your Photos library to an external disk but this is not convenient for obvious reasons as you wouldn't have access to your photos without the disk being mounted.

**Comment by anton-belkin** [comment]
I think it's very feasible to keep exporting while using Optimise Storage. I think I am going to do just that.
I want to have exported backup of my whole library on NAS, but do not want to continue to keep Originals on my mac.
The plan is following:
1) export all photos from your library using osxphotos
2) turn-on Optimise Storage
3) import some new photos over the course of say 3 months.
4) use osxphotos to export photos added within last 3 months, while downloading missing ones.
(the above assumes you have enough space to download 3 moths of photos)
I would also recommend using --update option while exporting overlapping time periods to make sure you have no gaps.

**Reply by RhetTbull** [reply]
Ideally osxphotos works best if your have download originals to this Mac enabled. However, you could do as @antonbelkinlykke said and download a batch of photos at a time using the --download-missing option. You could combine this with --limit to download only a batch at a time then give Photos time to re-optimize the library. You could also take a look at iCloud Photos Downloader https://github.com/icloud-photos-downloader/icloud_photos_downloader

**Comment by rajscode** [comment]
@synox A possibility to consider all via one Mac: 

Library A - iCloud connected and synced as optimized (reduced storage) version on internal drive. 
Library B - iCloud connected and synced as "Download Originals to this Mac" version to a USB or Thunderbolt connected external drive (SSD preferable)

Ensure the external drive is APFS formated.  That will enable you to also use osxphotos to export from Library B to a seperate directory structure on the external USB.  Storage space wise it won't use any more space because of APFS capablities to store the file once while providing the OS and apps a view it is stored multiple places on the same drive. 

Advantage of having osxphotos export to a seperate directory is that you'll also get each item independent of the Photos Library / iCloud sync made available to what ever backup app / script you have ... 

If that seems like something that could work -- see full details on how to implement at the following post:  [Full Backup Strategy - iCloud Full Photo Library on External Drive whilst optimized (reduced) version on Internal Drive](https://forums.macrumors.com/threads/backup-strategy-optimized-icloud-photo-library-on-mac-and-full-one-on-external-ssd.2351777/?post=31738760#post-31738760)
